chainer/variable.py

```diff
@@ -1088,158 +1088,232 @@ def T(self):
 
         """
-        self._to_device(device, allow_unchaining=False)
-
-    def _to_device(self, device, allow_unchaining):
         device = chainer.get_device(device)
 
         was_chainerx = self._has_chainerx_array
         is_chainerx = device.xp is chainerx
 
-        if not allow_unchaining:
-            if was_chainerx and not is_chainerx:
-                chx_arr = self._data[0]
-                if chx_arr is not None and chx_arr.is_backprop_required():
-                    raise RuntimeError(
-                        'A variable of a ChainerX array which requires '
-                        'gradients cannot be copied into non-chainerx device '
-                        '({}).'.format(device))
-            elif not was_chainerx and is_chainerx:
-                arr = self._data[0]
-                if arr is not None and self.creator is not None:
-                    raise RuntimeError(
-                        'A variable of a non-ChainerX array which is '
-                        'connected to a graph cannot be copied to a ChainerX '
-                        'device ({}).'.format(device))
-
         arr = self._data[0]
-        grad_var = self.grad_var
+
+        if not allow_unchain:
+            # `allow_unchain` is either `None` or `False`.
+            if was_chainerx:
+                is_unchain_required = (
+                    arr is not None
+                    and arr.is_backprop_required()
+                    and arr._is_chained())
+            else:
+                is_unchain_required = self.creator is not None
+
+            if is_unchain_required:
+                warnings.warn(
+                    'Device transfer that does not pass allow_unchain as True'
+                    ' for a variable that is connected to a graph will result'
+                    ' in an error in the future. It is recommended that you'
+                    ' pass allow_unchain=True if you do not care about any'
+                    ' existing graphs after the device transfer, or that you'
+                    ' unchain the graph beforehand if one exists.',
+                    DeprecationWarning)
 
         if was_chainerx and not is_chainerx:
             self._clear_chainerx()
             self._node = VariableNode(self, self._chainerx_name)
         elif not was_chainerx and is_chainerx:
             self._chainerx_name = self._node.name
 
-        self._device = device
-        self._has_chainerx_array = is_chainerx
+        # How the graph is unchained.
+        #
+        # - Non-ChainerX -> Non-ChainerX: Calling `unchain`.
+        # - Non-ChainerX -> ChainerX: Calling `unchain`.
+        # - ChainerX -> Non-ChainerX: Implicitly unchained.
+        # - ChainerX -> ChainerX: Calling `as_grad_stopped`.
 
-        if arr is None:
-            return
+        if not was_chainerx:
+            self.unchain()
 
-        if backend.get_device_from_array(arr) == device:
-            return
+        # Transfer gradient variable if any.
```

Comment:
Hmm, could you explain why the code is restructured this way (send grad_var -> send data -> set grad_var)?

Reply:
It's because `_set_chainerx_array` also modifies the gradient and passing the gradient to this method is rather tricky during device transfer (that's why it's done outside this method). IMO, this method can actually be simplified not to take the gradient argument at all.

Reply:
I'd like you to explain this part a bit more:
>passing the gradient to this method is rather tricky during device transfer

Reply:
Sorry let me rephrase it. I'm not sure how to do that in a clean way that's simpler than the current implementation. Do you have any ideas?

Reply:
I'm not saying which is better.
I'm simply asking you to explain the reason of the change.

Reply:
We now have to unchain `chainerx.ndarrays`, which we didn't have to before. This was basically the only way I could make it work. Internal methods such as `_set_chainerx_array`, `{_to,to}_device`, `grad_var` custom setter, etc. all highly depend on the internal state of the variable. Modifying the state of the variable without breaking it is non-trivial.

Reply:
Thanks, I have been asking that kind of information.

I don't still understand what you meant by `_set_chainerx_array` depends on the internal state. In my understanding the method is meant to be just a dumb assigner. The internal state should be handled by the caller.

Reply:
It may be dumb but it depends on `self. _requires_grad` and modifies `self._has_chainerx_array` which has a huge effect on the behavior of the variable as most attribute custom getters checks this variable.

Reply:
OK. I think the method should accept `requires_grad` as an agrument (in fact it was originally implemented that way). Do you think that could make the code simpler?

Reply:
That a might be a better design choice of `_has_chainerx_array` (it'll also be more clean when `grad` isn't used) but that alone won't help much for this PR (I'd like to use `Variable.to_device` for the gradient too, and not this internal method, since it's simpler). I think it's better to reconsider it at a Variable-wide scope if it should be refactored.

Reply:
I'm beginning to understand.

>I'd like to use `Variable.to_device` for the gradient too

This was actually the most important point of the refactoring and the change of the flow, right?
It would have been nice if you had revealed the intent at the very beginning.

What if `_set_chainerx_array` have `grad_var` as an argument?

Reply:
Perhaps I can merge it in the current status and consider refactoring later.

Reply:
I'm definitely for refactoring the Variable internals for the sake of maintainability and also to do it separately outside this PR. It should have more time and though put into it.

Reply:
OK, I agree.